Drug stock api

harimohanraj89 opened the pull request:
**Story card:** [sc-7687](https://app.shortcut.com/simpledotorg/story/7687/drug-stock-api)

This PR introduces a `/api/v4/drug_stocks` API endpoint for the mobile app to query in order to power the drug stock reminder nudges in the app.

* Accepts requests for a specific month, and uses the current user's facility (X-Facility-ID header)
* Returns a 200 with drug stock info if a drug stock report has been submitted for that facility and month
* Returns a 400 of no report has been submitted yet

### To Do

* [x] Fix the API specs and update documentation

prabhanshuguptagit commented on app/controllers/api/v4/drug_stocks_controller.rb:
```diff
@@ -0,0 +1,20 @@
+  private
+
+  def for_end_of_month
+    @for_end_of_month ||= Date.parse(params[:date]).end_of_month
```
`Date.parse` accepts a wide range of date formats as valid input. Let's use `Date.strptime` and specify the exact format.

prabhanshuguptagit commented on app/controllers/api/v4/drug_stocks_controller.rb:
```diff
@@ -0,0 +1,20 @@
+
+  def for_end_of_month
+    @for_end_of_month ||= Date.parse(params[:date]).end_of_month
+  rescue Date::Error, TypeError
```
Is the `TypeError` to rescue non-string params being passed? Should we coerce `params[:date]` to a string first in that case?

harimohanraj89 replied:
The TypeError is to catch `nil` - ie, date was not provided.

prabhanshuguptagit commented on app/controllers/api/v4/drug_stocks_controller.rb:
```diff
@@ -0,0 +1,20 @@
+class Api::V4::DrugStocksController < APIController
+  def index
+    if for_end_of_month.nil?
+      render status: 400, json: {errors: ["valid `date` must be provided - eg. 2021-10-29"]}
```
Shall we also specify the date format we expect in this (already very helpful) error message?

prabhanshuguptagit commented on spec/controllers/api/v4/drug_stocks_controller_spec.rb:
```diff
@@ -0,0 +1,77 @@
+
+        expect(response.status).to eq(400)
+        expect(response.body).to include("date")
+      end
```
👌🏼

prabhanshuguptagit commented on spec/controllers/api/v4/drug_stocks_controller_spec.rb:
```diff
@@ -0,0 +1,77 @@
+        end
+      }
+
+      allow(DrugStock).to receive(:latest_for_facilities).with([facility], end_of_month).and_return(drug_stocks)
```
Is this mocking needed if the drug stocks have been created in the right month at `facility`?

harimohanraj89 replied:
Ah it is indeed not. I will remove.

prabhanshuguptagit reviewed (approved):
Nifty :shipit:

harimohanraj89 commented:
Thank you for the review, addressed feedback in [9f892ab](https://github.com/simpledotorg/simple-server/pull/3635/commits/9f892ab2f74fc47774923863d41084ddc24b1571) 👍
